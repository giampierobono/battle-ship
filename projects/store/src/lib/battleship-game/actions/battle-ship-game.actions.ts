import { createAction, props } from '@ngrx/store';
import { PlayerFiresPositionPayloadModel } from '../../players';

export const tryFirePosition = createAction('[Game] Player fire!', props<PlayerFiresPositionPayloadModel>());
