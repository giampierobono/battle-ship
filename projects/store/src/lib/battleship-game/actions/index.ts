export * from './battle-ship-game.actions';
