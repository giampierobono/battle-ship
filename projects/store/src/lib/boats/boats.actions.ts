import { createAction, props } from '@ngrx/store';
import { AddBoatToPlayerPayload, HitPlayerBoatPayloadModel } from './models';

export const addBoatToPlayer = createAction('[Boats] Add boat to player', props<AddBoatToPlayerPayload>());

export const hitPlayerBoat = createAction('[Boats] Hit player boat', props<HitPlayerBoatPayloadModel>());
