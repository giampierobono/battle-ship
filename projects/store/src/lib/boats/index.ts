export * from './boats.actions';
export * from './boats.reducer';
export * from './models';
