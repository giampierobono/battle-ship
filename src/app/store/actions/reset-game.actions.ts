import { createAction } from '@ngrx/store';

export const resetGame = createAction('[BS webapp] Reset game');
