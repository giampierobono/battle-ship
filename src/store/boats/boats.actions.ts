import { createAction, props } from '@ngrx/store';
import { AddBoatsToPlayerPayload, HitPlayerBoatPayloadModel } from './models';

export const addBoatsToPlayer = createAction('[Boats] Add boat to player', props<AddBoatsToPlayerPayload>());

export const hitPlayerBoat = createAction('[Boats] Hit player boat', props<HitPlayerBoatPayloadModel>());
