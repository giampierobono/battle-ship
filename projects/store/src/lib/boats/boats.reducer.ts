import { BoatModel } from 'models';
import { createReducer } from '@ngrx/store';
import { produceOn } from 'store-tools';
import * as BoatsActions from './boats.actions';

export interface IBoatsState {
  boats: BoatModel[];
}

export const boatsInitialState: IBoatsState = {
  boats: [],
};

export const boatsReducer = createReducer(
  boatsInitialState,
  produceOn(BoatsActions.addBoatToPlayer, (state, { boat }) => {
    state.boats.push(boat);
  }),
  produceOn(BoatsActions.addBoatsToPlayer, (state, { boats }) => {
    state.boats.push(...boats);
  }),
  produceOn(BoatsActions.hitPlayerBoat, (state, { position }) => {
    state.boats.forEach((boat) => {
      if (
        boat.positions[position.row][position.column] !== undefined &&
        boat.positions[position.row][position.column] !== true
      ) {
        boat.positions[position.row][position.column] = true;
        boat.hits++;
      }
    });
  })
);
