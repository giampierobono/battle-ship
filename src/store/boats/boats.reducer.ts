import { createReducer } from '@ngrx/store';
import * as BoatsActions from './boats.actions';
import { BoatModel, PositionStateEnum } from '../../models';
import { produceOn } from '../common/utils';

export interface IBoatsState {
  boats: Array<Array<BoatModel>>;
}

export const boatsInitialState: IBoatsState = {
  boats: [],
};

export const boatsReducer = createReducer(
  boatsInitialState,
  produceOn(BoatsActions.addBoatsToPlayer, (state, { boats }) => {
    state.boats = boats;
  }),
  produceOn(BoatsActions.hitPlayerBoat, (state, { position: { playerIndex, row, column }, boatIndex }) => {
    state.boats[playerIndex][boatIndex].positions[row][column] = PositionStateEnum.HIT;
    state.boats[playerIndex][boatIndex].hits++;
  })
);
