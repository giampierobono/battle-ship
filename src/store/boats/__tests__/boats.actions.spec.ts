import { addBoatsToPlayer, hitPlayerBoat } from '../boats.actions';

describe('Boats actions', () => {
  describe('addBoatsToPlayer', () => {
    it('should return correct action structure', () => {
      expect(
        addBoatsToPlayer({ boats: [[{ boatIndex: 0, playerIndex: 0, hits: 0, shapeKey: 2, positions: {} }]] })
      ).toMatchSnapshot();
    });
  });

  describe('hitPlayerBoat', () => {
    it('should return correct action structure', () => {
      expect(hitPlayerBoat({ position: { playerIndex: 0, row: 0, column: 0 }, boatIndex: 0 })).toMatchSnapshot();
    });
  });
});
