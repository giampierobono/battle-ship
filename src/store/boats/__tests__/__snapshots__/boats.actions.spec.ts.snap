// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Boats actions addBoatsToPlayer should return correct action structure 1`] = `
Object {
  "boats": Array [
    Array [
      Object {
        "boatIndex": 0,
        "hits": 0,
        "playerIndex": 0,
        "positions": Object {},
        "shapeKey": 2,
      },
    ],
  ],
  "type": "[Boats] Add boat to player",
}
`;

exports[`Boats actions hitPlayerBoat should return correct action structure 1`] = `
Object {
  "boatIndex": 0,
  "position": Object {
    "column": 0,
    "playerIndex": 0,
    "row": 0,
  },
  "type": "[Boats] Hit player boat",
}
`;
